src/passes/HeapStoreOptimization.cpp

```diff
@@ -271,6 +321,114 @@ struct HeapStoreOptimization
+  //    (struct.set (local.get $x) (..br $out..))  ;; X' here has a br
+  //  )
+  //  ..use $x..
+  //
+  // Note how we do the local.set first. Imagine we optimized to this:
+  //
+  //  (block $out
+  //    (local.set $x (struct.new (..br $out..) Y Z))
+  //  )
+  //  ..use $x..
+  //
+  // Now the br happens first, skipping the local.set entirely, and the use
+  // later down will not get the proper value.
+  //
+  // To check for this problem, see which basic blocks can be reached from the
+  // value. The structured IR limits what is possible here, since we know that
+  // the struct.set follows the local.set: the only problem is one like we
+  // showed in the example, where we branch to something after the
+  // struct.set's basic block, in post-order. As we are processing the blocks
+  // in that same order, we just need to see that we cannot reach any block we
+  // have yet to see.
+  //
+  // Note that there may be more basic blocks after the struct.set, but they
+  // cannot be reached due to the structured control flow:
+  //
+  //  (block $out
+  //    (local.set ..
+  //    (struct.set ..
+  //    (if
+  //      ..more control flow..
+  //    )
+  //  )
+  //
+  // At this point in the traversal we have reached the block, so we've seen
+  // the if's basic blocks. But nothing in the struct.set's value can branch
+  // to them: there is no way to jump into the middle of a block.
+  bool canSkipLocalSet(StructSet* set) {
+    auto* blockBeforeValue = expressionBlocks[set->ref];
+    if (!blockBeforeValue) {
+      // We are in unreachable code. Return that we are doing something
+      // dangerous, as this is not worth optimizing.
+      return true;
+    }
+    // We will stop the flow when we reach the struct.set itself: it is fine for
+    // control flow to get there, that is what normally happens.
+    auto* structSetBlock = expressionBlocks[set];
+    if (!structSetBlock) {
+      // We are in unreachable code.
+      return true;
+    }
+
+    if (blockBeforeValue == structSetBlock) {
+      // There is no control flow at all in the area we are concerned with, so
+      // there is no way to skip the local.set.
+      return false;
+    }
+
+    // Given those blocks, we can perform the following analysis: see if we can
+    // get out of the range of blocks [blockBeforeValue .. structSetBlock].
+    // Normal execution reaches structSetBlock if there is no branching, so that
+    // is what we expect; any branch further forward without going through that
+    // block is suspect. Likewise, any branch backwards to before the value is
+    // also suspect as it may reach a loop header (and there may be a local.get
+    // usage there).
+    //
+    // This is not entirely precise, as we do not search for actually dangerous
+    // local.gets, that is, this has false positives. But it is a simple
+    // analysis that works in almost all cases, and avoids potentially large
+    // amounts of flow work.
```

Comment:
This seems like a good opportunity to use the new lazy local graph utility. Not only could the results become more precise, but I'd bet that this code would become simpler as well since we wouldn't have to do any new flowing.

Reply:
I don't believe a flow can be avoided here, or at least I can't see how. E.g.
```wat
(block $out
  (local.set $x (struct.new X Y Z))
  (struct.set (local.get $x) (..br $out..)) ;; this local.get is ok
  (call $foo (local.get $x))                ;; this local.get is ok
)
(call $bar (local.get $x))                  ;; this local.get is not
```
If we use a LocalGraph to find the uses of that local.set, we'd see three: the first is in the `struct.set`, which is expected. But the other two need to be differentiated: the former is ok since we never reach it if we branch away and skip the `struct.set`, but the latter is dangerous in the optimized form, which is this:
```wat
(block $out
  (local.set $x (struct.new (..br $out..) Y Z))
  (call $foo (local.get $x))
)
(call $bar (local.get $x))
```
(now we may skip the `local.set`, and read a null at the bottom).

The one way I can see to maybe do it is to construct a LocalGraph both before and after the optimization. Then, computing the sets for the `local.get` on the call to `$foo` would detect a change: before we had just one possible set, but after we'd have two or more (whatever value is possible before the `local.set`). However, even with laziness, creating a new graph for the state after, each time we try to optimize an expression, seems like more work.

Reply:
Hmm, I guess the problem is that local graph doesn't calculate liveness, exactly. Will take another look at this code shortly.

Reply:
I experimented with this in a side branch and I have something that seems to work. It adjusts the IR for the state after the optimization and then runs a LazyLocalGraph to see what happens with the gets of the `local.set`. I think the logic I figured out there is right, it passes tests, but I didn't fuzz it.

The code is not super-short but it may be simpler than this PR. However, I think speed is a bigger problem. This has to construct a new LazyLocalGraph for each individual `struct.set` we see a possible branching in the value of, since we modify the IR and look at that, and the modification is different for each one. Even a LazyLocalGraph is not cheap as it does scan the function the first time it is called upon - linear time, but doing it for each `struct.set` can be quadratic. I'm not sure I see a way around that, though I am fairly sure this situation - where we see possible branching - is pretty rare, so it might be ok in practice...

([unpolished code](https://github.com/WebAssembly/binaryen/compare/main...kripken:binaryen:effect.throws2.b.c.d?expand=1))